nmigen/back/rtlil.py

```diff
@@ -639,7 +636,7 @@ def on_Signal(self, value):
         return wire_next or wire_curr
 
     def _prepare_value_for_Slice(self, value):
-        assert isinstance(value, (ast.Signal, ast.Slice, ast.Cat, rec.Record))
+        assert isinstance(value, (ast.Signal, ast.Slice, ast.Cat, ast.UserValue))
```

Comment:
This change is actually quite subtle. Since `UserValue`s cannot by themselves be translated, and should be lowered first, this assertion would not work correctly (and neither would the other implementation of `_prepare_value_for_Slice`). I think what you need to do here is to lower the `UserValue` in `on_Slice` *before* passing it to `_prepare_value_for_Slice`.

Reply:
okay

Reply:
do we want to allow returning other UserValues from the lower method?
If i get it right this is currently not supported but if we do not allow it now we would break all code that lowers UserValues into Records

Reply:
Good question! That's something I did not consider. I think that for now, we can change `UserValue._lazy_lower` to recursively call `_lazy_lower` again if it gets another `UserValue`. Once #355 is solved, this logic should be a part of `Value.cast`.

Reply:
okay i will do that too in the rtlil backend

Reply:
You should call `_lazy_lower` in the RTLIL backend, since that is the nMigen-side API of `UserValue`s.

Reply:
bikeshed: isn't _lazy_lower() is not really lazy, its more caching?

Reply:
The "laziness" in case of `_lazy_lower` is because you can use a `UserValue` without committing to a specific shape upfront, only doing that on first use. The caching aspect is incidental, it is there to guarantee correctness (internal consistency of nMigen data structures).